src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs

```diff
@@ -55,57 +61,202 @@ internal static partial class JsonWriterHelper
+        private static readonly Vector128<sbyte> s_mask_SByte_0x60 = Vector128.Create((sbyte)0x60); // Grave Access '`'
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Vector128<short> CreateEscapingMask(Vector128<short> sourceValue)
         {
-            int idx;
+            Debug.Assert(Sse2.IsSupported);
 
-            if (encoder != null)
-            {
-                idx = encoder.FindFirstCharacterToEncodeUtf8(value);
-                goto Return;
-            }
+            Vector128<short> mask = Sse2.CompareLessThan(sourceValue, s_mask_UInt16_0x20); // Space ' ', anything in the control characters range
+
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_UInt16_0x22)); // Quotation Mark '"'
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_UInt16_0x26)); // Ampersand '&'
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_UInt16_0x27)); // Apostrophe '''
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_UInt16_0x2B)); // Plus sign '+'
+
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_UInt16_0x3C)); // Less Than Sign '<'
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_UInt16_0x3E)); // Greater Than Sign '>'
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_UInt16_0x5C)); // Reverse Solidus '\'
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_UInt16_0x60)); // Grave Access '`'
+
+            mask = Sse2.Or(mask, Sse2.CompareGreaterThan(sourceValue, s_mask_UInt16_0x7E)); // Tilde '~', anything above the ASCII range
+
+            return mask;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Vector128<sbyte> CreateEscapingMask(Vector128<sbyte> sourceValue)
+        {
+            Debug.Assert(Sse2.IsSupported);
 
-            for (idx = 0; idx < value.Length; idx++)
+            Vector128<sbyte> mask = Sse2.CompareLessThan(sourceValue, s_mask_SByte_0x20); // Control characters, and anything above 0x7E since sbyte.MaxValue is 0x7E
+
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_SByte_0x22)); // Quotation Mark "
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_SByte_0x26)); // Ampersand &
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_SByte_0x27)); // Apostrophe '
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_SByte_0x2B)); // Plus sign +
+
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_SByte_0x3C)); // Less Than Sign <
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_SByte_0x3E)); // Greater Than Sign >
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_SByte_0x5C)); // Reverse Solidus \
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_SByte_0x60)); // Grave Access `
```

Comment:
>> There are quite a lot comparisons and ors (hence also a quite long raw-chain).
I have an approach that uses bit-masks, need to verify performance (later today).

@gfoidl we actually have a hard-coded bitmask already that we use for ASCII. Yes using that instead of `Sse2Or` seems to be the best solution.

Reply:
The bitmask I refer to is https://github.com/gfoidl/corefx/blob/5b5f8c6a8ecb95dca9559c40e3f1cdc483ed45e2/src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs#L156-L173

> we use for ASCII

Can you point me to that, please? So far I didn't find one that matches with my idea.
(Anyway, I'm preparing a PR as @ahsonkhan suggested, well see then there.).

Reply:
> we actually have a hard-coded bitmask already that we use for ASCII

> I wonder what the break-even point is?

I shared the perf comparison of "current" (which uses the hard-coded bit-mask for ASCII vs "alternatives/this PR" in the PR description:
https://gist.github.com/ahsonkhan/c566f5e7d65c1fde5a83a67be290c4ee

Using intrinsics is always faster (so you could think of a break-even point as 8 which is when we start using intrinsics). We do have a ~10-15% regression for strings <= 7 characters because of the extra length check (but in absolute terms that's in the 1 ns range).

Reply:
> That's not true. We get substantial perf wins using Sse2 operations. Keep in mind, all these operations are not happening per character, but per 8 characters.

I was referring to @gfoidl `NeedsEscapingSsse3` example which is appears faster than the SSe2 (and was not referring to the current implementation that doesn't use SSe2). Perhaps SSS3 isn't as common as SSe2 so we should use SSe2 instead (or use SSs3 if available).

Reply:
`SSE3` was introduced in 2004 and `SSSE3` was introduced in 2006. It is reasonable to expect that the vast majority of users running .NET Core would have this support and it is therefore beneficial to include a code-path for it.

Reply:
Makes sense @tannergooding

@steveharter - gotcha. And agreed. I will let @gfoidl drive that change and do the perf testing to evaluate its feasibility and we can discuss its merit over that PR. I want to remove this code from S.T.Json first though (especially for porting to 3.1), so the optimization and the escaping logic can solely live in S.T.E.Web.

Reply:
With https://github.com/dotnet/corefx/pull/42023 I'm having a look how this bitmask-approach can be incorporated in S.T.E.Web, and not in S.T.Json. If not possible I'll target S.T.Json.

Reply:
> I'm having a look how this bitmask-approach can be incorporated in S.T.E.Web, and not in S.T.Json

Sounds good. Look at the concrete types that implement `JavaScriptEncoder` since those are what will get used more often.

I'd say the following would be most common (with `TextEncoder` or a custom `JavascriptEncoder` being quite uncommon, not worth optimizing):
- `JavaScriptEncoder.Default` - used by default
- `JavaScriptEncoder.UnsafeRelaxedJsonEscaping` - used by default within aspnet
- `JavaScriptEncoder.Create(UnicodeRanges.All)` - used by folks who want non-ascii text left unescaped

> If not possible I'll target S.T.Json.

I would prefer not to have encoding specific logic duplicated in S.T.Json, so let's try to avoid that if possible (let me know if you get blocked because of this for some reason, because my assumption is whatever was feasible in S.T.Json is also feasible in S.T.E.W).

Reply:
Here we go: https://github.com/dotnet/corefx/pull/42073 :wink: